Chisel plugin bundle elements handler (#2306)
chipsalliance/chisel3: docs/src/appendix/chisel-plugin.md

# Notes on the Compiler Plug-in

The Chisel plugin provides some operations that are too difficult, or not possbile, 
to implement through regular Scala code.

## Compiler plugin operations
1. Automatically generates the `cloneType` methods of Bundles
2. Records that this plugin ran on a bundle by adding a method `_usingPlugin: Boolean = true`
3. Generates a function that returns list of the hardware fields in a bundle `_elementsImpl`

> 

Reviewer: I think this is useful information, but it feels like a weird mix of "useful to users" and "useful only to contributors". I think it would be better to either split this into 2 files, or at least call out with a `##` level title ("Notes for Contributors") or something.

Reviewer 2: Agreed -- for example should this appear on the website?

Reviewer: I'd generally say the contributor only stuff should not live on the website, but I'm not to fussed either way, it just needs to be clear what is what.

Author: @jackkoenig @mwachs5 I agree this is weird. I moved this purposely into the README instead of placing it where it would be picked up by website. This should be a short term option that quickly (maybe by 3.5.2) becomes the default.
I tried to simplify and condense things a bit. Not sure how much more to do here

> 

Reviewer 2: 4. Future work -- detecting Seq[Data] in Bundles? I'm confused about what that section is saying

Author: This section was just enumerating the things the plugin currently does.
I've added a bit more verbiage to this effect.

### Generating `cloneType` method

> After a period of testing this is now on by default. This flag was used to control that prior to full adoption.

Reviewer 2: ```suggestion
After a period of testing this is now on by default (as of Chisel Foo.Baz.Biddle). This flag was used to control that prior to full adoption.
```

Author: Referenced date and pr where current behavior was implemented

```
-P:chiselplugin:useBundlePlugin
```

### Generating the `_elementsImpl` method
For generating the elementsImpl method, the default is off.

> Turn it on by using the javac flag

Reviewer: ```suggestion
Turn it on by using the scalac flag (scalacOptions in SBT)
```

Author: will do

```
-P:chiselplugin:buildElementAccessor
```

> #### Detecting Bundles with Seq[Data]

Reviewer 2: I'm confused what this section is saying. does the compiler plugin detect this or not? What does "currently" mean?

Author: I've cleaned up the language a bit here. Basically says this feature (3. on earlier list) is off by default and uses a scalac flag to turn it on.

> Currently having this is a run time error
> Here is a block of code that could be added (with some refinement in 
> the detection mechanism)

Reviewer 2: ```suggestion

Trying to have a `val Seq[Data]` (as opposed to a `val Vec[Data]` in a `Bundle` is
a run time error.
Here is a block of code that could be added to the plugin to detect this case at compile time (with some refinement in 
the detection mechanism):
```

Author: will add this to next commit on this PR

```scala
  //TODO: The following code block creates a compiler error for bundle fields that are
  //      Seq[Data]. The test for this case is a bit too hacky so for now we will just
  //      elide these fields, and will be caught in Aggregate.scala at runtime
  if (member.isAccessor && typeIsSeqOfData(member.tpe) && !isIgnoreSeqInBundle(bundleSymbol)) {
    global.reporter.error(
      member.pos,
      s"Bundle.field ${bundleSymbol.name}.${member.name} cannot be a Seq[Data]. " +
        "Use Vec or MixedVec or mix in trait IgnoreSeqInBundle"
    )
  }
```

> ## Notes about working on the `_elementsImpl` generator for the plugin in `BundleComponent.scala`

Reviewer 2: ```suggestion
### Notes about working on the `_elementsImpl` generator for the plugin in `BundleComponent.scala`
```

Author: done

> # Notes on the Compiler Plug-in

Reviewer 2: you need to do an editing pass on this file actually looking at it in a markdown viewer, many issues in `.md` format

Author: I made this a bit cleaner, it's not too thorough but has some better debugging hints. Used IntelliJ's md rendering.
I hope that is sufficient.

The Chisel plugin provides some features that are too difficult, or not possbile, 

> to implement through regular code. These features

Reviewer 2: Something is missing in this sentence

## Compiler plugin Features

> - Records that this plugin ran on a bundle by adding a method `_usingPlugin: Boolean = true`
> - Automatically generating the `cloneType` methods of Bundles
> - Generating a function that returns list of the hardware fields in a bundle

Reviewer 2: ```suggestion
- Records that this plugin ran on a bundle by adding a method `_usingPlugin: Boolean = true`
- Automatically generates the `cloneType` methods of Bundles
- Generates a function that returns a list of the hardware fields in a bundle
```

Author: This change will be in next commit
